### PR TITLE
# New multi-currency exchange rate editor

### PR DESCRIPTION
**Pull request by atm-jpb**
## Context
Currently multicurrency logs past exchange rates when you "update" the exchange rate of a currency, but it doesn't enable you to correct past exchange rates or to date an exchange rate when you add it (the date is automatically set to the current time).

## This PR
This PR introduces a new page that enables you to view, edit or add exchange rates in a way somewhat similar to Dolibarr dictionaries. The page is accessible via the administration module.

image

### CODE REVIEW

**Comment by atm-jpb** [comment]
@eldy  : have some error message from stickler-ci "Could not fetch pull request branch". i have no clue how to fix this one. 
have you any direction ?

**Comment by atm-jpb** [comment]
@eldy hello  i come back to you  with  my message from stickler-ci "Could not fetch pull request branch".

**Comment by eldy** [comment]
You can ignore stickler-ci errors. It is not important.
However, the errors into Travis CI are important. Fixing them is mandatory.

**Comment by atm-jpb** [comment]
@eldy conflict solved

**Comment by atm-jpb** [comment]
@eldy if i understand the PR process , this one is ok to your human review ?

File: htdocs/multicurrency/multicurrency_rate.php [hunk 1]
```diff
@@ -0,0 +1,547 @@
...
+if (! $sortfield) $sortfield="cr.date_sync";
+if (! $sortorder) $sortorder="ASC";
+
+
+// Initialize technical object to manage hooks. Note that conf->hooks_modules contains array of hooks
+$object=new CurrencyRate($db);
+
+$extrafields = new ExtraFields($db);
+$form=new Form($db);
+
```
Comment (atm-john) [inline htdocs/multicurrency/multicurrency_rate.php:75]: Missing $hookmanager->initHooks

File: htdocs/multicurrency/multicurrency_rate.php [hunk 2]
```diff
@@ -0,0 +1,547 @@
...
+if ($action == 'update'){
+	$currencyRate = new CurrencyRate($db);
+	$result = $currencyRate->fetch($id_rate_selected);
+	if ( $result > 0){
+		$currency_static  = new MultiCurrency($db);
+		$fk_currency = $currency_static->getIdFromCode($db, $multicurrency_code);
+		$currencyRate->date_sync = $db->escape(GETPOST('dateinput'));
+		$currencyRate->fk_multicurrency = $fk_currency;
+		$currencyRate->rate = $db->escape(GETPOST('rateinput'));
```
Comment (atm-john) [inline htdocs/multicurrency/multicurrency_rate.php:132]: Your GETPOST is not typed for number ?

File: htdocs/multicurrency/multicurrency_rate.php [hunk 3]
```diff
@@ -0,0 +1,547 @@
...
+	}
+}
+
+if ($action == 'update'){
+	$currencyRate = new CurrencyRate($db);
+	$result = $currencyRate->fetch($id_rate_selected);
+	if ( $result > 0){
+		$currency_static  = new MultiCurrency($db);
+		$fk_currency = $currency_static->getIdFromCode($db, $multicurrency_code);
+		$currencyRate->date_sync = $db->escape(GETPOST('dateinput'));
```
Comment (atm-john) [inline htdocs/multicurrency/multicurrency_rate.php:130]: you don't check the format : keep this in your mind : never trust user input

File: htdocs/multicurrency/multicurrency_rate.php [hunk 4]
```diff
@@ -0,0 +1,547 @@
...
+
+if ($action == "confirm_delete"){
+	$current_rate = new CurrencyRate($db);
+	$current_rate->fetch(intval($id_rate_selected));
+	if ($current_rate){
+		$result  = $current_rate->delete();
+		if ($result){
+			setEventMessage($langs->trans('successRateDelete'));
+		}else {
+			setEventMessage($langs->trans('errorRateDelete'));
```
Comment (atm-john) [inline htdocs/multicurrency/multicurrency_rate.php:177]: not an error ?

File: htdocs/multicurrency/multicurrency_rate.php [hunk 5]
```diff
@@ -0,0 +1,547 @@
...
+	print '<table><tr>';
+
+	print ' <td>' . $langs->trans('date') . '</td>';
+	print ' <td><input class="minwidth200" name="dateinput" value="' . dol_escape_htmltag($dateinput) . '" type="date"></td>';
+
+	print ' <td>' . $langs->trans('Codemulticurrency') . '</td>';
+	print '<td>' . $form->selectMultiCurrency((GETPOSTISSET('multicurrency_code') ? GETPOST('multicurrency_code', 'alpha') : $multicurrency_code), 'multicurrency_code', 0, " code != '".$conf->currency."'", true) . '</td>';
+
+	print ' <td>' . $langs->trans('rate') . '</td>';
+	print ' <td><input class="minwidth200" name="rateinput" value="' . dol_escape_htmltag($rateinput) . '" type="text"></td>';
```
Comment (atm-john) [inline htdocs/multicurrency/multicurrency_rate.php:252]: use type number with min = 0 and step = any

**Pull request merged**